src/node/routes/vscode.ts

```diff
@@ -33,27 +34,43 @@ export class CodeServerRouteWrapper {
       })
     }
 
-    const { query } = await req.settings.read()
-    if (query) {
-      // Ew means the workspace was closed so clear the last folder/workspace.
-      if (req.query.ew) {
-        delete query.folder
-        delete query.workspace
-      }
+    let folder = undefined
+    let workspace = undefined
+    const to = self(req)
+    const settings = await req.settings.read()
+    const lastOpened = settings.query || {}
+
+    // Ew means the workspace was closed so clear the last folder/workspace.
+    if (req.query.ew) {
+      delete lastOpened.folder
+      delete lastOpened.workspace
+    }
 
+    if (!req.query.folder && !req.query.workspace) {
       // Redirect to the last folder/workspace if nothing else is opened.
       if (
-        !req.query.folder &&
-        !req.query.workspace &&
-        (query.folder || query.workspace) &&
+        (lastOpened.folder || lastOpened.workspace) &&
         !req.args["ignore-last-opened"] // This flag disables this behavior.
       ) {
-        const to = self(req)
         return redirect(req, res, to, {
-          folder: query.folder,
-          workspace: query.workspace,
+          folder: lastOpened.folder,
+          workspace: lastOpened.workspace,
         })
```

Comment:
We could do this to share the redirect block below:

```suggestion
        folder = lastOpened.folder
        workspace = lastOpened.workspace
```